store/src/java/com/zimbra/cs/service/account/GetHAB.java

```diff
@@ -0,0 +1,152 @@
+        String rootGrpId = request.getAttribute(AccountConstants.A_HAB_ROOT_GROUP_ID);
+        Account acct = getAuthenticatedAccount(zsc);
+
+        ZimbraLog.account.info("Request for HAB group:%s by: ", rootGrpId, acct);
+
+        Map<String, HABGroup> groups = new HashMap<String, HABGroup>();
+        List<HABGroup> childGrpList = new ArrayList<HABGroup>();
+        LdapDistributionList group = (LdapDistributionList) prov.getGroupBasic(Key.DistributionListBy.id, rootGrpId);
+        Set<String> members = group.getAllMembersSet();
+        
+        HABGroup grp = new HABGroup();
+        grp.setId(group.getId());
+        grp.setName(group.getName());
+        grp.setRootGroup(ZmBoolean.TRUE);
+        groups.put(group.getMail(), grp);
+        grp.setAttrs(group.getAttrs());
+
+        for (String member : members) {
+            HABGroup grpChild = new HABGroup();
+            grpChild.setParentGroupId(group.getId());
+            groups.put(member, grpChild);
+            childGrpList.add(grpChild);
+        }
+        Collections.sort(childGrpList , new SortByName());
+        grp.setChildGroups(childGrpList);
+
+        List<LdapDistributionList> lists = prov
+            .getAllHabGroups(prov.get(DomainBy.name, acct.getDomainName()));
+
+        for (LdapDistributionList list : lists) {
+            String key = list.getAttr(Provisioning.A_mail);
+            HABGroup habGrp = groups.get(key);
+            if (isUnderRoot(group.getDN(), list.getDN())) {
+                if (habGrp == null) {
+                    habGrp = new HABGroup();
+                }
+                if (habGrp.getRootGroup().compareTo(ZmBoolean.TRUE) == 0) {
+                    continue;
+                }
+                habGrp.setId(list.getId());
+                habGrp.setName(list.getName());
+                habGrp.setAttrs(list.getAttrs());
+                groups.put(key, habGrp);
+                List<HABGroup> children = new ArrayList<HABGroup>();
+                for (String member : list.getAllMembersSet()) {
+                    HABGroup grpChild = groups.get(member);
+                    if (grpChild == null) {
+                        grpChild = new HABGroup();
+                    }
+                    grpChild.setParentGroupId(list.getId());
+                    groups.put(member, grpChild);
+                    children.add(grpChild);
+                }
+                Collections.sort(children , new SortByName());
+                habGrp.setChildGroups(children);
+            }
+        }
+
+        Element response = zsc.createElement(AccountConstants.E_GET_HAB_RESPONSE);
+        ToXML.encodeHabGroup(response, grp, null);
+
+        return response;
+    }
+    
+    /**
+     * @param attr
+     * @param dn
+     * @return
+     */
+    private boolean isUnderRoot(String rootGroupDn, String targetGroupDn) {
+       return targetGroupDn.contains(rootGroupDn);
+    }
+    
+    class SortByName implements Comparator<HABGroup> {
```

Comment:
should sort by SeniorityIndex first. if seniorityIndex is not set, those groups will get default seniorityIndex as 0, and then sort them as per name.